packages/core/core/generate-text/generate-text.ts

```diff
@@ -108,80 +112,173 @@ case of misconfigured tools.
 By default, it's set to 0, which will disable the feature.
      */
     maxToolRoundtrips?: number;
-  }): Promise<GenerateTextResult<TOOLS>> {
-  const retry = retryWithExponentialBackoff({ maxRetries });
-  const validatedPrompt = getValidatedPrompt({ system, prompt, messages });
 
-  const mode = {
-    type: 'regular' as const,
-    ...prepareToolsAndToolChoice({ tools, toolChoice }),
+    /**
+     * Optional telemetry information.
+     */
+    // This is meant to be both flexible for custom app requirements (metadata)
+    // and extensible for standardization (example: functionId, more to come).
+    experimental_telemetry?: {
+      /**
+       * Enable or disable telemetry. Disabled by default.
+       */
+      isEnabled?: boolean;
+
+      functionId?: string;
+      metadata?: Record<string, AttributeValue>;
+    };
+  }): Promise<GenerateTextResult<TOOLS>> {
+  const baseTelemetryAttributes = {
+    'ai.model.provider': model.provider,
+    'ai.model.id': model.modelId,
+    // settings:
+    'ai.settings.maxRetries': maxRetries,
+    ...Object.entries(settings ?? {}).reduce((attributes, [key, value]) => {
+      attributes[`ai.settings.${key}`] = value;
+      return attributes;
+    }, {} as Record<string, AttributeValue>),
+    // special telemetry information
+    'operation.name': 'ai.generateText',
+    'resource.name': telemetry?.functionId,
```

Comment:
Ideally there should be some value for `resource.name`. Is there any good default when `telemetry?.functionId` is falsy?

Reply:
not really - it's a stateless function call. The arguments (prompt) are variable, so even hashing would not work, and the model is likely used in other places as well. I can set it to `"unknown"` but all those solutions mean that it'll be hard to distinguish different generateText locations without a functionId.

Reply:
Ok. Let's keep undefined then.